Index: cmake/modules/LLVM-Config.cmake
===================================================================
--- cmake/modules/LLVM-Config.cmake
+++ cmake/modules/LLVM-Config.cmake
@@ -9,26 +9,30 @@
 endfunction()
 
 
-function(is_llvm_target_library library return_var)
-  # Sets variable `return_var' to ON if `library' corresponds to a
-  # LLVM supported target. To OFF if it doesn't.
+function(is_omitted_target_lib library return_var)
+  # Sets variable `return_var' to ON if `library' corresponds to an omitted LLVM
+  # target. To OFF otherwise.
   set(${return_var} OFF PARENT_SCOPE)
   string(TOUPPER "${library}" capitalized_lib)
-  string(TOUPPER "${LLVM_ALL_TARGETS}" targets)
+  set(omitted_targets ${LLVM_ALL_TARGETS})
+  list(REMOVE_ITEM omitted_targets ${LLVM_TARGETS_TO_BUILD})
+  string(TOUPPER "${omitted_targets}" targets)
   foreach(t ${targets})
     if( capitalized_lib STREQUAL t OR
-        capitalized_lib STREQUAL "LLVM${t}" OR
-        capitalized_lib STREQUAL "LLVM${t}CODEGEN" OR
-        capitalized_lib STREQUAL "LLVM${t}ASMPARSER" OR
-        capitalized_lib STREQUAL "LLVM${t}ASMPRINTER" OR
-        capitalized_lib STREQUAL "LLVM${t}DISASSEMBLER" OR
-        capitalized_lib STREQUAL "LLVM${t}INFO" )
+        capitalized_lib STREQUAL "${t}" OR
+        capitalized_lib STREQUAL "${t}DESC" OR
+        capitalized_lib STREQUAL "${t}CODEGEN" OR
+        capitalized_lib STREQUAL "${t}ASMPARSER" OR
+        capitalized_lib STREQUAL "${t}ASMPRINTER" OR
+        capitalized_lib STREQUAL "${t}DISASSEMBLER" OR
+        capitalized_lib STREQUAL "${t}INFO" OR
+        capitalized_lib STREQUAL "${t}UTILS" OR
+        capitalized_lib STREQUAL "${t}INSTPRINTER" )
       set(${return_var} ON PARENT_SCOPE)
       break()
     endif()
   endforeach()
-endfunction(is_llvm_target_library)
-
+endfunction(is_omitted_target_lib)
 
 macro(llvm_config executable)
   cmake_parse_arguments(ARG "USE_SHARED" "" "" ${ARGN})
@@ -132,15 +136,24 @@
       if( TARGET LLVM${c}AsmParser )
         list(APPEND expanded_components "LLVM${c}AsmParser")
       endif()
+      if( TARGET LLVM${c}AsmParser )
+        list(APPEND expanded_components "LLVM${c}CodeGen")
+      endif()
       if( TARGET LLVM${c}Desc )
         list(APPEND expanded_components "LLVM${c}Desc")
       endif()
       if( TARGET LLVM${c}Info )
         list(APPEND expanded_components "LLVM${c}Info")
       endif()
+      if( TARGET LLVM${c}AsmParser )
+        list(APPEND expanded_components "LLVM${c}InstPrinter")
+      endif()
       if( TARGET LLVM${c}Disassembler )
         list(APPEND expanded_components "LLVM${c}Disassembler")
       endif()
+      if( TARGET LLVM${c}AsmParser )
+        list(APPEND expanded_components "LLVM${c}Utils")
+      endif()
     elseif( c STREQUAL "native" )
       # already processed
     elseif( c STREQUAL "nativecodegen" )
@@ -197,10 +210,19 @@
       string(TOUPPER "${c}" capitalized)
       list(FIND capitalized_libs LLVM${capitalized} lib_idx)
       if( lib_idx LESS 0 )
-        # The component is unknown. Maybe is an omitted target?
-        is_llvm_target_library(${c} iltl_result)
-        if( NOT iltl_result )
+        # The component is neither a known target (X86, MSP430, etc.) nor a
+        # component that has been scanned. Check if it's part of an omitted
+        # target, which is the only case when error is certain.
+        is_omitted_target_lib(${c} omitted_result)
+        if(omitted_result)
           message(FATAL_ERROR "Library `${c}' not found in list of llvm libraries.")
+        else()
+          # ${c} is either a malformed component, or one that is yet to be
+          # scanned. Ideally, we could distinguish between the two by scanning
+          # components in RPO of their dependency graph. But this isn't the
+          # case, so defer the well-formedness check to CMake at generation
+          # time.
+          list(APPEND expanded_components "$<LINK_ONLY:LLVM${c}>")
         endif()
       else( lib_idx LESS 0 )
         list(GET LLVM_AVAILABLE_LIBS ${lib_idx} canonical_lib)
Index: lib/Analysis/CMakeLists.txt
===================================================================
--- lib/Analysis/CMakeLists.txt
+++ lib/Analysis/CMakeLists.txt
@@ -86,4 +86,10 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Core
+    Support
+    ProfileData
+    Object
+)
Index: lib/AsmParser/CMakeLists.txt
===================================================================
--- lib/AsmParser/CMakeLists.txt
+++ lib/AsmParser/CMakeLists.txt
@@ -9,4 +9,8 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Core
+    Support
+)
Index: lib/Bitcode/Reader/CMakeLists.txt
===================================================================
--- lib/Bitcode/Reader/CMakeLists.txt
+++ lib/Bitcode/Reader/CMakeLists.txt
@@ -10,4 +10,8 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Core
+    Support
+)
Index: lib/Bitcode/Writer/CMakeLists.txt
===================================================================
--- lib/Bitcode/Writer/CMakeLists.txt
+++ lib/Bitcode/Writer/CMakeLists.txt
@@ -6,4 +6,9 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    Core
+    Support
+)
Index: lib/CodeGen/AsmPrinter/CMakeLists.txt
===================================================================
--- lib/CodeGen/AsmPrinter/CMakeLists.txt
+++ lib/CodeGen/AsmPrinter/CMakeLists.txt
@@ -25,4 +25,15 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    CodeGen
+    Core
+    DebugInfoCodeView
+    DebugInfoMSF
+    MC
+    MCParser
+    Support
+    Target
+)
Index: lib/CodeGen/CMakeLists.txt
===================================================================
--- lib/CodeGen/CMakeLists.txt
+++ lib/CodeGen/CMakeLists.txt
@@ -151,7 +151,18 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    BitReader
+    BitWriter
+    Core
+    MC
+    ScalarOpts
+    Support
+    Target
+    TransformUtils
+)
 
 add_subdirectory(SelectionDAG)
 add_subdirectory(AsmPrinter)
Index: lib/CodeGen/GlobalISel/CMakeLists.txt
===================================================================
--- lib/CodeGen/GlobalISel/CMakeLists.txt
+++ lib/CodeGen/GlobalISel/CMakeLists.txt
@@ -32,4 +32,13 @@
         
         DEPENDS
         intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    CodeGen
+    Core
+    MC
+    Support
+    Target
+    TransformUtils
+)
Index: lib/CodeGen/MIRParser/CMakeLists.txt
===================================================================
--- lib/CodeGen/MIRParser/CMakeLists.txt
+++ lib/CodeGen/MIRParser/CMakeLists.txt
@@ -5,4 +5,12 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    AsmParser
+    CodeGen
+    Core
+    MC
+    Support
+    Target
+)
Index: lib/CodeGen/SelectionDAG/CMakeLists.txt
===================================================================
--- lib/CodeGen/SelectionDAG/CMakeLists.txt
+++ lib/CodeGen/SelectionDAG/CMakeLists.txt
@@ -26,4 +26,13 @@
   
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    CodeGen
+    Core
+    MC
+    Support
+    Target
+    TransformUtils
+)
Index: lib/DebugInfo/CodeView/CMakeLists.txt
===================================================================
--- lib/DebugInfo/CodeView/CMakeLists.txt
+++ lib/DebugInfo/CodeView/CMakeLists.txt
@@ -18,4 +18,8 @@
 
   ADDITIONAL_HEADER_DIRS
   ${LLVM_MAIN_INCLUDE_DIR}/llvm/DebugInfo/CodeView
-  )
+
+  LINK_COMPONENTS
+    Support
+    DebugInfoMSF
+)
Index: lib/DebugInfo/DWARF/CMakeLists.txt
===================================================================
--- lib/DebugInfo/DWARF/CMakeLists.txt
+++ lib/DebugInfo/DWARF/CMakeLists.txt
@@ -24,4 +24,8 @@
   ADDITIONAL_HEADER_DIRS
   ${LLVM_MAIN_INCLUDE_DIR}/llvm/DebugInfo/DWARF
   ${LLVM_MAIN_INCLUDE_DIR}/llvm/DebugInfo
-  )
+
+  LINK_COMPONENTS
+    Object
+    Support
+)
Index: lib/DebugInfo/MSF/CMakeLists.txt
===================================================================
--- lib/DebugInfo/MSF/CMakeLists.txt
+++ lib/DebugInfo/MSF/CMakeLists.txt
@@ -7,4 +7,7 @@
   StreamWriter.cpp
   ADDITIONAL_HEADER_DIRS
   "${LLVM_MAIN_INCLUDE_DIR}/llvm/DebugInfo/MSF"
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/DebugInfo/PDB/CMakeLists.txt
===================================================================
--- lib/DebugInfo/PDB/CMakeLists.txt
+++ lib/DebugInfo/PDB/CMakeLists.txt
@@ -98,6 +98,12 @@
 
   ADDITIONAL_HEADER_DIRS
   ${LIBPDB_ADDITIONAL_HEADER_DIRS}
-  )
+
+  LINK_COMPONENTS
+    Object
+    Support
+    DebugInfoCodeView
+    DebugInfoMSF
+)
 
 target_link_libraries(LLVMDebugInfoPDB INTERFACE "${LIBPDB_ADDITIONAL_LIBRARIES}")
Index: lib/DebugInfo/Symbolize/CMakeLists.txt
===================================================================
--- lib/DebugInfo/Symbolize/CMakeLists.txt
+++ lib/DebugInfo/Symbolize/CMakeLists.txt
@@ -5,4 +5,10 @@
 
   ADDITIONAL_HEADER_DIRS
   ${LLVM_MAIN_INCLUDE_DIR}/llvm/DebugInfo/Symbolize
-  )
+
+  LINK_COMPONENTS
+    DebugInfoDWARF
+    DebugInfoPDB
+    Object
+    Support
+)
Index: lib/Demangle/CMakeLists.txt
===================================================================
--- lib/Demangle/CMakeLists.txt
+++ lib/Demangle/CMakeLists.txt
@@ -1,3 +1,5 @@
 add_llvm_library(LLVMDemangle
   ItaniumDemangle.cpp
+
+  LINK_COMPONENTS
 )
Index: lib/ExecutionEngine/CMakeLists.txt
===================================================================
--- lib/ExecutionEngine/CMakeLists.txt
+++ lib/ExecutionEngine/CMakeLists.txt
@@ -12,7 +12,15 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Core
+    MC
+    Object
+    RuntimeDyld
+    Support
+    Target
+)
 
 add_subdirectory(Interpreter)
 add_subdirectory(MCJIT)
Index: lib/ExecutionEngine/IntelJITEvents/CMakeLists.txt
===================================================================
--- lib/ExecutionEngine/IntelJITEvents/CMakeLists.txt
+++ lib/ExecutionEngine/IntelJITEvents/CMakeLists.txt
@@ -12,6 +12,14 @@
   jitprofiling.c
 
   LINK_LIBS ${LLVM_INTEL_JIT_LIBS}
+
+  LINK_COMPONENTS
+    CodeGen
+    Core
+    DebugInfoDWARF
+    Support
+    Object
+    ExecutionEngine
 )
 
 add_dependencies(LLVMIntelJITEvents LLVMCodeGen)
Index: lib/ExecutionEngine/Interpreter/CMakeLists.txt
===================================================================
--- lib/ExecutionEngine/Interpreter/CMakeLists.txt
+++ lib/ExecutionEngine/Interpreter/CMakeLists.txt
@@ -13,7 +13,13 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    CodeGen
+    Core
+    ExecutionEngine
+    Support
+)
 
 if( LLVM_ENABLE_FFI )
   target_link_libraries( LLVMInterpreter PRIVATE ${FFI_LIBRARY_PATH} )
Index: lib/ExecutionEngine/MCJIT/CMakeLists.txt
===================================================================
--- lib/ExecutionEngine/MCJIT/CMakeLists.txt
+++ lib/ExecutionEngine/MCJIT/CMakeLists.txt
@@ -3,4 +3,12 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Core
+    ExecutionEngine
+    Object
+    RuntimeDyld
+    Support
+    Target
+)
Index: lib/ExecutionEngine/OProfileJIT/CMakeLists.txt
===================================================================
--- lib/ExecutionEngine/OProfileJIT/CMakeLists.txt
+++ lib/ExecutionEngine/OProfileJIT/CMakeLists.txt
@@ -4,4 +4,9 @@
 add_llvm_library(LLVMOProfileJIT
   OProfileJITEventListener.cpp
   OProfileWrapper.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+    Object
+    ExecutionEngine
+)
Index: lib/ExecutionEngine/Orc/CMakeLists.txt
===================================================================
--- lib/ExecutionEngine/Orc/CMakeLists.txt
+++ lib/ExecutionEngine/Orc/CMakeLists.txt
@@ -12,4 +12,12 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Core
+    ExecutionEngine
+    Object
+    RuntimeDyld
+    Support
+    TransformUtils
+)
Index: lib/ExecutionEngine/RuntimeDyld/CMakeLists.txt
===================================================================
--- lib/ExecutionEngine/RuntimeDyld/CMakeLists.txt
+++ lib/ExecutionEngine/RuntimeDyld/CMakeLists.txt
@@ -10,4 +10,9 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    MC
+    Object
+    Support
+)
Index: lib/IR/CMakeLists.txt
===================================================================
--- lib/IR/CMakeLists.txt
+++ lib/IR/CMakeLists.txt
@@ -59,4 +59,7 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/IRReader/CMakeLists.txt
===================================================================
--- lib/IRReader/CMakeLists.txt
+++ lib/IRReader/CMakeLists.txt
@@ -6,4 +6,10 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    AsmParser
+    BitReader
+    Core
+    Support
+)
Index: lib/LTO/CMakeLists.txt
===================================================================
--- lib/LTO/CMakeLists.txt
+++ lib/LTO/CMakeLists.txt
@@ -62,4 +62,22 @@
 
   DEPENDS
   intrinsics_gen
+
+  LINK_COMPONENTS
+    Analysis
+    BitReader
+    BitWriter
+    CodeGen
+    Core
+    ipo
+    InstCombine
+    Linker
+    MC
+    ObjCARCOpts
+    Object
+    Passes
+    ScalarOpts
+    Support
+    Target
+    TransformUtils
 )
Index: lib/LibDriver/CMakeLists.txt
===================================================================
--- lib/LibDriver/CMakeLists.txt
+++ lib/LibDriver/CMakeLists.txt
@@ -4,5 +4,10 @@
 
 add_llvm_library(LLVMLibDriver
   LibDriver.cpp
-  )
+
+  LINK_COMPONENTS
+    Object
+    Option
+    Support
+)
 add_dependencies(LLVMLibDriver LibOptionsTableGen)
Index: lib/LineEditor/CMakeLists.txt
===================================================================
--- lib/LineEditor/CMakeLists.txt
+++ lib/LineEditor/CMakeLists.txt
@@ -9,6 +9,8 @@
   ${LLVM_MAIN_INCLUDE_DIR}/llvm/LineEditor
 
   LINK_LIBS
-  LLVMSupport
   ${link_libs}
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Linker/CMakeLists.txt
===================================================================
--- lib/Linker/CMakeLists.txt
+++ lib/Linker/CMakeLists.txt
@@ -7,4 +7,9 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Core
+    Support
+    TransformUtils
+)
Index: lib/MC/CMakeLists.txt
===================================================================
--- lib/MC/CMakeLists.txt
+++ lib/MC/CMakeLists.txt
@@ -50,7 +50,10 @@
 
   ADDITIONAL_HEADER_DIRS
   ${LLVM_MAIN_INCLUDE_DIR}/llvm/MC
-  )
+
+  LINK_COMPONENTS
+    Support
+)
 
 add_subdirectory(MCParser)
 add_subdirectory(MCDisassembler)
Index: lib/MC/MCDisassembler/CMakeLists.txt
===================================================================
--- lib/MC/MCDisassembler/CMakeLists.txt
+++ lib/MC/MCDisassembler/CMakeLists.txt
@@ -4,4 +4,8 @@
   MCExternalSymbolizer.cpp
   MCRelocationInfo.cpp
   MCSymbolizer.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+)
Index: lib/MC/MCParser/CMakeLists.txt
===================================================================
--- lib/MC/MCParser/CMakeLists.txt
+++ lib/MC/MCParser/CMakeLists.txt
@@ -11,4 +11,8 @@
 
   ADDITIONAL_HEADER_DIRS
   ${LLVM_MAIN_INCLUDE_DIR}/llvm/MC/MCParser
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+)
Index: lib/Object/CMakeLists.txt
===================================================================
--- lib/Object/CMakeLists.txt
+++ lib/Object/CMakeLists.txt
@@ -23,4 +23,11 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    BitReader
+    Core
+    MC
+    MCParser
+    Support
+)
Index: lib/ObjectYAML/CMakeLists.txt
===================================================================
--- lib/ObjectYAML/CMakeLists.txt
+++ lib/ObjectYAML/CMakeLists.txt
@@ -5,4 +5,7 @@
   MachOYAML.cpp
   ObjectYAML.cpp
   DWARFYAML.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Option/CMakeLists.txt
===================================================================
--- lib/Option/CMakeLists.txt
+++ lib/Option/CMakeLists.txt
@@ -6,4 +6,7 @@
 
   ADDITIONAL_HEADER_DIRS
   ${LLVM_MAIN_INCLUDE_DIR}/llvm/Option
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Passes/CMakeLists.txt
===================================================================
--- lib/Passes/CMakeLists.txt
+++ lib/Passes/CMakeLists.txt
@@ -6,4 +6,16 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    CodeGen
+    Core
+    ipo
+    InstCombine
+    ScalarOpts
+    Support
+    TransformUtils
+    Vectorize
+    Instrumentation
+)
Index: lib/ProfileData/CMakeLists.txt
===================================================================
--- lib/ProfileData/CMakeLists.txt
+++ lib/ProfileData/CMakeLists.txt
@@ -12,6 +12,10 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Core
+    Support
+)
 
 add_subdirectory(Coverage)
Index: lib/ProfileData/Coverage/CMakeLists.txt
===================================================================
--- lib/ProfileData/Coverage/CMakeLists.txt
+++ lib/ProfileData/Coverage/CMakeLists.txt
@@ -8,4 +8,10 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Core
+    Object
+    ProfileData
+    Support
+)
Index: lib/Support/CMakeLists.txt
===================================================================
--- lib/Support/CMakeLists.txt
+++ lib/Support/CMakeLists.txt
@@ -136,6 +136,9 @@
   ${LLVM_MAIN_INCLUDE_DIR}/llvm/Support
 
   LINK_LIBS ${system_libs}
-  )
+
+  LINK_COMPONENTS
+    Demangle
+)
 
 set_property(TARGET LLVMSupport PROPERTY LLVM_SYSTEM_LIBS "${system_libs}")
Index: lib/TableGen/CMakeLists.txt
===================================================================
--- lib/TableGen/CMakeLists.txt
+++ lib/TableGen/CMakeLists.txt
@@ -10,4 +10,7 @@
 
   ADDITIONAL_HEADER_DIRS
   ${LLVM_MAIN_INCLUDE_DIR}/llvm/TableGen
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/AArch64/AsmParser/CMakeLists.txt
===================================================================
--- lib/Target/AArch64/AsmParser/CMakeLists.txt
+++ lib/Target/AArch64/AsmParser/CMakeLists.txt
@@ -2,5 +2,13 @@
 
 add_llvm_library(LLVMAArch64AsmParser
   AArch64AsmParser.cpp
-  )
+
+  LINK_COMPONENTS
+    AArch64Desc
+    AArch64Info
+    AArch64Utils
+    MC
+    MCParser
+    Support
+)
 
Index: lib/Target/AArch64/CMakeLists.txt
===================================================================
--- lib/Target/AArch64/CMakeLists.txt
+++ lib/Target/AArch64/CMakeLists.txt
@@ -70,6 +70,22 @@
 
   DEPENDS
   intrinsics_gen
+
+  LINK_COMPONENTS
+    AArch64AsmPrinter
+    AArch64Desc
+    AArch64Info
+    AArch64Utils
+    Analysis
+    AsmPrinter
+    CodeGen
+    Core
+    MC
+    ScalarOpts
+    SelectionDAG
+    Support
+    Target
+    GlobalISel
 )
 
 add_subdirectory(TargetInfo)
Index: lib/Target/AArch64/Disassembler/CMakeLists.txt
===================================================================
--- lib/Target/AArch64/Disassembler/CMakeLists.txt
+++ lib/Target/AArch64/Disassembler/CMakeLists.txt
@@ -3,6 +3,14 @@
 add_llvm_library(LLVMAArch64Disassembler
   AArch64Disassembler.cpp
   AArch64ExternalSymbolizer.cpp
-  )
+
+  LINK_COMPONENTS
+    AArch64Desc
+    AArch64Info
+    AArch64Utils
+    MC
+    MCDisassembler
+    Support
+)
 
 add_dependencies(LLVMAArch64Disassembler AArch64CommonTableGen)
Index: lib/Target/AArch64/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/AArch64/InstPrinter/CMakeLists.txt
+++ lib/Target/AArch64/InstPrinter/CMakeLists.txt
@@ -2,6 +2,11 @@
 
 add_llvm_library(LLVMAArch64AsmPrinter
   AArch64InstPrinter.cpp
-  )
+
+  LINK_COMPONENTS
+    AArch64Utils
+    MC
+    Support
+)
 
 add_dependencies(LLVMAArch64AsmPrinter AArch64CommonTableGen)
Index: lib/Target/AArch64/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/AArch64/MCTargetDesc/CMakeLists.txt
+++ lib/Target/AArch64/MCTargetDesc/CMakeLists.txt
@@ -8,6 +8,12 @@
   AArch64MCTargetDesc.cpp
   AArch64MachObjectWriter.cpp
   AArch64TargetStreamer.cpp
+
+  LINK_COMPONENTS
+    AArch64AsmPrinter
+    AArch64Info
+    MC
+    Support
 )
 add_dependencies(LLVMAArch64Desc AArch64CommonTableGen)
 
Index: lib/Target/AArch64/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/AArch64/TargetInfo/CMakeLists.txt
+++ lib/Target/AArch64/TargetInfo/CMakeLists.txt
@@ -2,6 +2,9 @@
 
 add_llvm_library(LLVMAArch64Info
   AArch64TargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
 
 add_dependencies(LLVMAArch64Info AArch64CommonTableGen)
Index: lib/Target/AArch64/Utils/CMakeLists.txt
===================================================================
--- lib/Target/AArch64/Utils/CMakeLists.txt
+++ lib/Target/AArch64/Utils/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMAArch64Utils
   AArch64BaseInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/AMDGPU/AsmParser/CMakeLists.txt
===================================================================
--- lib/Target/AMDGPU/AsmParser/CMakeLists.txt
+++ lib/Target/AMDGPU/AsmParser/CMakeLists.txt
@@ -1,5 +1,13 @@
 add_llvm_library(LLVMAMDGPUAsmParser
   AMDGPUAsmParser.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    MCParser
+    AMDGPUDesc
+    AMDGPUInfo
+    AMDGPUUtils
+    Support
+)
 
 add_dependencies(LLVMAMDGPUAsmParser LLVMAMDGPUUtils)
Index: lib/Target/AMDGPU/CMakeLists.txt
===================================================================
--- lib/Target/AMDGPU/CMakeLists.txt
+++ lib/Target/AMDGPU/CMakeLists.txt
@@ -85,7 +85,25 @@
   SITypeRewriter.cpp
   SIWholeQuadMode.cpp
   ${GLOBAL_ISEL_BUILD_FILES}
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    AsmPrinter
+    CodeGen
+    Core
+    ipo
+    MC
+    AMDGPUAsmPrinter
+    AMDGPUDesc
+    AMDGPUInfo
+    AMDGPUUtils
+    ScalarOpts
+    SelectionDAG
+    Support
+    Target
+    TransformUtils
+    Vectorize
+)
 
 add_subdirectory(AsmParser)
 add_subdirectory(InstPrinter)
Index: lib/Target/AMDGPU/Disassembler/CMakeLists.txt
===================================================================
--- lib/Target/AMDGPU/Disassembler/CMakeLists.txt
+++ lib/Target/AMDGPU/Disassembler/CMakeLists.txt
@@ -2,6 +2,14 @@
 
 add_llvm_library(LLVMAMDGPUDisassembler
   AMDGPUDisassembler.cpp
-  )
+
+  LINK_COMPONENTS
+    AMDGPUDesc
+    AMDGPUInfo
+    AMDGPUUtils
+    MC
+    MCDisassembler
+    Support
+)
 
 add_dependencies(LLVMAMDGPUDisassembler AMDGPUCommonTableGen LLVMAMDGPUUtils)
Index: lib/Target/AMDGPU/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/AMDGPU/InstPrinter/CMakeLists.txt
+++ lib/Target/AMDGPU/InstPrinter/CMakeLists.txt
@@ -1,5 +1,10 @@
 add_llvm_library(LLVMAMDGPUAsmPrinter
   AMDGPUInstPrinter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+    AMDGPUUtils
+)
 
 add_dependencies(LLVMAMDGPUAsmPrinter LLVMAMDGPUUtils)
Index: lib/Target/AMDGPU/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/AMDGPU/MCTargetDesc/CMakeLists.txt
+++ lib/Target/AMDGPU/MCTargetDesc/CMakeLists.txt
@@ -10,4 +10,12 @@
   AMDGPUTargetStreamer.cpp
   R600MCCodeEmitter.cpp
   SIMCCodeEmitter.cpp
-  )
+
+  LINK_COMPONENTS
+    Core
+    MC
+    AMDGPUAsmPrinter
+    AMDGPUInfo
+    AMDGPUUtils
+    Support
+)
Index: lib/Target/AMDGPU/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/AMDGPU/TargetInfo/CMakeLists.txt
+++ lib/Target/AMDGPU/TargetInfo/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMAMDGPUInfo
   AMDGPUTargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/AMDGPU/Utils/CMakeLists.txt
===================================================================
--- lib/Target/AMDGPU/Utils/CMakeLists.txt
+++ lib/Target/AMDGPU/Utils/CMakeLists.txt
@@ -2,4 +2,9 @@
   AMDGPUBaseInfo.cpp
   AMDKernelCodeTUtils.cpp
   AMDGPUAsmUtils.cpp
-  )
+
+  LINK_COMPONENTS
+    Core
+    MC
+    Support
+)
Index: lib/Target/ARM/AsmParser/CMakeLists.txt
===================================================================
--- lib/Target/ARM/AsmParser/CMakeLists.txt
+++ lib/Target/ARM/AsmParser/CMakeLists.txt
@@ -1,3 +1,10 @@
 add_llvm_library(LLVMARMAsmParser
   ARMAsmParser.cpp
-  )
+
+  LINK_COMPONENTS
+    ARMDesc
+    ARMInfo
+    MC
+    MCParser
+    Support
+)
Index: lib/Target/ARM/CMakeLists.txt
===================================================================
--- lib/Target/ARM/CMakeLists.txt
+++ lib/Target/ARM/CMakeLists.txt
@@ -61,7 +61,22 @@
   Thumb2SizeReduction.cpp
   ARMComputeBlockSize.cpp
   ${GLOBAL_ISEL_BUILD_FILES}
-  )
+
+  LINK_COMPONENTS
+    ARMAsmPrinter
+    ARMDesc
+    ARMInfo
+    Analysis
+    AsmPrinter
+    CodeGen
+    Core
+    MC
+    ScalarOpts
+    SelectionDAG
+    Support
+    Target
+    GlobalISel
+)
 
 add_subdirectory(TargetInfo)
 add_subdirectory(AsmParser)
Index: lib/Target/ARM/Disassembler/CMakeLists.txt
===================================================================
--- lib/Target/ARM/Disassembler/CMakeLists.txt
+++ lib/Target/ARM/Disassembler/CMakeLists.txt
@@ -1,3 +1,9 @@
 add_llvm_library(LLVMARMDisassembler
   ARMDisassembler.cpp
-  )
+
+  LINK_COMPONENTS
+    ARMDesc
+    ARMInfo
+    MCDisassembler
+    Support
+)
Index: lib/Target/ARM/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/ARM/InstPrinter/CMakeLists.txt
+++ lib/Target/ARM/InstPrinter/CMakeLists.txt
@@ -1,3 +1,7 @@
 add_llvm_library(LLVMARMAsmPrinter
   ARMInstPrinter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+)
Index: lib/Target/ARM/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/ARM/MCTargetDesc/CMakeLists.txt
+++ lib/Target/ARM/MCTargetDesc/CMakeLists.txt
@@ -13,4 +13,11 @@
   ARMUnwindOpAsm.cpp
   ARMWinCOFFObjectWriter.cpp
   ARMWinCOFFStreamer.cpp
-  )
+
+  LINK_COMPONENTS
+    ARMAsmPrinter
+    ARMInfo
+    MC
+    MCDisassembler
+    Support
+)
Index: lib/Target/ARM/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/ARM/TargetInfo/CMakeLists.txt
+++ lib/Target/ARM/TargetInfo/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMARMInfo
   ARMTargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/AVR/AsmParser/CMakeLists.txt
===================================================================
--- lib/Target/AVR/AsmParser/CMakeLists.txt
+++ lib/Target/AVR/AsmParser/CMakeLists.txt
@@ -1,3 +1,10 @@
 add_llvm_library(LLVMAVRAsmParser
   AVRAsmParser.cpp
+
+  LINK_COMPONENTS
+    MC
+    MCParser
+    AVRDesc
+    AVRInfo
+    Support
 )
Index: lib/Target/AVR/CMakeLists.txt
===================================================================
--- lib/Target/AVR/CMakeLists.txt
+++ lib/Target/AVR/CMakeLists.txt
@@ -33,6 +33,18 @@
 
   DEPENDS
   intrinsics_gen
+
+  LINK_COMPONENTS
+    AsmPrinter
+    CodeGen
+    Core
+    MC
+    AVRAsmPrinter
+    AVRDesc
+    AVRInfo
+    SelectionDAG
+    Support
+    Target
 )
 
 add_subdirectory(AsmParser)
Index: lib/Target/AVR/Disassembler/CMakeLists.txt
===================================================================
--- lib/Target/AVR/Disassembler/CMakeLists.txt
+++ lib/Target/AVR/Disassembler/CMakeLists.txt
@@ -1,4 +1,9 @@
 add_llvm_library(LLVMAVRDisassembler
   AVRDisassembler.cpp
+
+  LINK_COMPONENTS
+    MCDisassembler
+    AVRInfo
+    Support
 )
 
Index: lib/Target/AVR/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/AVR/InstPrinter/CMakeLists.txt
+++ lib/Target/AVR/InstPrinter/CMakeLists.txt
@@ -3,6 +3,10 @@
 
 add_llvm_library(LLVMAVRAsmPrinter
   AVRInstPrinter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+)
 
 add_dependencies(LLVMAVRAsmPrinter AVRCommonTableGen)
\ No newline at end of file
Index: lib/Target/AVR/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/AVR/MCTargetDesc/CMakeLists.txt
+++ lib/Target/AVR/MCTargetDesc/CMakeLists.txt
@@ -7,6 +7,12 @@
   AVRMCExpr.cpp
   AVRMCTargetDesc.cpp
   AVRTargetStreamer.cpp
+
+  LINK_COMPONENTS
+    MC
+    AVRAsmPrinter
+    AVRInfo
+    Support
 )
 
 add_dependencies(LLVMAVRDesc AVRCommonTableGen)
Index: lib/Target/AVR/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/AVR/TargetInfo/CMakeLists.txt
+++ lib/Target/AVR/TargetInfo/CMakeLists.txt
@@ -3,6 +3,10 @@
 
 add_llvm_library(LLVMAVRInfo
   AVRTargetInfo.cpp
+
+  LINK_COMPONENTS
+    MC
+    Support
 )
 
 add_dependencies(LLVMAVRInfo AVRCommonTableGen)
Index: lib/Target/BPF/CMakeLists.txt
===================================================================
--- lib/Target/BPF/CMakeLists.txt
+++ lib/Target/BPF/CMakeLists.txt
@@ -21,7 +21,19 @@
   BPFRegisterInfo.cpp
   BPFSubtarget.cpp
   BPFTargetMachine.cpp
-  )
+
+  LINK_COMPONENTS
+    AsmPrinter
+    CodeGen
+    Core
+    MC
+    BPFAsmPrinter
+    BPFDesc
+    BPFInfo
+    SelectionDAG
+    Support
+    Target
+)
 
 add_subdirectory(Disassembler)
 add_subdirectory(InstPrinter)
Index: lib/Target/BPF/Disassembler/CMakeLists.txt
===================================================================
--- lib/Target/BPF/Disassembler/CMakeLists.txt
+++ lib/Target/BPF/Disassembler/CMakeLists.txt
@@ -1,4 +1,9 @@
 add_llvm_library(LLVMBPFDisassembler
   BPFDisassembler.cpp
+
+  LINK_COMPONENTS
+    MCDisassembler
+    BPFInfo
+    Support
 )
 
Index: lib/Target/BPF/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/BPF/InstPrinter/CMakeLists.txt
+++ lib/Target/BPF/InstPrinter/CMakeLists.txt
@@ -1,3 +1,7 @@
 add_llvm_library(LLVMBPFAsmPrinter
   BPFInstPrinter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+)
Index: lib/Target/BPF/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/BPF/MCTargetDesc/CMakeLists.txt
+++ lib/Target/BPF/MCTargetDesc/CMakeLists.txt
@@ -3,4 +3,10 @@
   BPFAsmBackend.cpp
   BPFMCCodeEmitter.cpp
   BPFELFObjectWriter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    BPFAsmPrinter
+    BPFInfo
+    Support
+)
Index: lib/Target/BPF/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/BPF/TargetInfo/CMakeLists.txt
+++ lib/Target/BPF/TargetInfo/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMBPFInfo
   BPFTargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/CMakeLists.txt
===================================================================
--- lib/Target/CMakeLists.txt
+++ lib/Target/CMakeLists.txt
@@ -11,7 +11,13 @@
 
   ADDITIONAL_HEADER_DIRS
   ${LLVM_MAIN_INCLUDE_DIR}/llvm/Target
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    Core
+    MC
+    Support
+)
 
 foreach(t ${LLVM_TARGETS_TO_BUILD})
   message(STATUS "Targeting ${t}")
Index: lib/Target/Hexagon/AsmParser/CMakeLists.txt
===================================================================
--- lib/Target/Hexagon/AsmParser/CMakeLists.txt
+++ lib/Target/Hexagon/AsmParser/CMakeLists.txt
@@ -2,6 +2,13 @@
 
 add_llvm_library(LLVMHexagonAsmParser
   HexagonAsmParser.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    MCParser
+    Support
+    HexagonDesc
+    HexagonInfo
+)
 
 add_dependencies( LLVMHexagonAsmParser HexagonCommonTableGen )
Index: lib/Target/Hexagon/CMakeLists.txt
===================================================================
--- lib/Target/Hexagon/CMakeLists.txt
+++ lib/Target/Hexagon/CMakeLists.txt
@@ -58,7 +58,22 @@
   RDFDeadCode.cpp
   RDFGraph.cpp
   RDFLiveness.cpp
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    AsmPrinter
+    CodeGen
+    Core
+    HexagonAsmParser
+    HexagonDesc
+    HexagonInfo
+    MC
+    ScalarOpts
+    SelectionDAG
+    Support
+    Target
+    TransformUtils
+)
 
 add_subdirectory(AsmParser)
 add_subdirectory(TargetInfo)
Index: lib/Target/Hexagon/Disassembler/CMakeLists.txt
===================================================================
--- lib/Target/Hexagon/Disassembler/CMakeLists.txt
+++ lib/Target/Hexagon/Disassembler/CMakeLists.txt
@@ -1,3 +1,10 @@
 add_llvm_library(LLVMHexagonDisassembler
   HexagonDisassembler.cpp
-  )
+
+  LINK_COMPONENTS
+    HexagonDesc
+    HexagonInfo
+    MC
+    MCDisassembler
+    Support
+)
Index: lib/Target/Hexagon/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/Hexagon/MCTargetDesc/CMakeLists.txt
+++ lib/Target/Hexagon/MCTargetDesc/CMakeLists.txt
@@ -13,6 +13,11 @@
   HexagonMCShuffler.cpp
   HexagonMCTargetDesc.cpp
   HexagonShuffler.cpp
-  )
+
+  LINK_COMPONENTS
+    HexagonInfo
+    MC
+    Support
+)
 
 add_dependencies(LLVMHexagonDesc HexagonCommonTableGen)
Index: lib/Target/Hexagon/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/Hexagon/TargetInfo/CMakeLists.txt
+++ lib/Target/Hexagon/TargetInfo/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMHexagonInfo
   HexagonTargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/Lanai/AsmParser/CMakeLists.txt
===================================================================
--- lib/Target/Lanai/AsmParser/CMakeLists.txt
+++ lib/Target/Lanai/AsmParser/CMakeLists.txt
@@ -2,6 +2,13 @@
 
 add_llvm_library(LLVMLanaiAsmParser
   LanaiAsmParser.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    MCParser
+    Support
+    LanaiDesc
+    LanaiInfo
+)
 
 add_dependencies( LLVMLanaiAsmParser LanaiCommonTableGen )
Index: lib/Target/Lanai/CMakeLists.txt
===================================================================
--- lib/Target/Lanai/CMakeLists.txt
+++ lib/Target/Lanai/CMakeLists.txt
@@ -26,6 +26,21 @@
   LanaiSubtarget.cpp
   LanaiTargetMachine.cpp
   LanaiTargetObjectFile.cpp
+
+  LINK_COMPONENTS
+    Analysis
+    AsmPrinter
+    CodeGen
+    Core
+    LanaiAsmParser
+    LanaiDesc
+    LanaiInfo
+    LanaiInstPrinter
+    MC
+    SelectionDAG
+    Support
+    Target
+    TransformUtils
 )
 
 add_subdirectory(AsmParser)
Index: lib/Target/Lanai/Disassembler/CMakeLists.txt
===================================================================
--- lib/Target/Lanai/Disassembler/CMakeLists.txt
+++ lib/Target/Lanai/Disassembler/CMakeLists.txt
@@ -1,3 +1,10 @@
 add_llvm_library(LLVMLanaiDisassembler
   LanaiDisassembler.cpp
-  )
+
+  LINK_COMPONENTS
+    LanaiDesc
+    LanaiInfo
+    MC
+    MCDisassembler
+    Support
+)
Index: lib/Target/Lanai/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/Lanai/InstPrinter/CMakeLists.txt
+++ lib/Target/Lanai/InstPrinter/CMakeLists.txt
@@ -1,3 +1,7 @@
 add_llvm_library(LLVMLanaiInstPrinter
   LanaiInstPrinter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+)
Index: lib/Target/Lanai/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/Lanai/MCTargetDesc/CMakeLists.txt
+++ lib/Target/Lanai/MCTargetDesc/CMakeLists.txt
@@ -5,4 +5,11 @@
   LanaiMCCodeEmitter.cpp
   LanaiMCExpr.cpp
   LanaiMCTargetDesc.cpp
+
+  LINK_COMPONENTS
+    LanaiInfo
+    LanaiInstPrinter
+    MC
+    MCDisassembler
+    Support
 )
Index: lib/Target/Lanai/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/Lanai/TargetInfo/CMakeLists.txt
+++ lib/Target/Lanai/TargetInfo/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMLanaiInfo
   LanaiTargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/MSP430/CMakeLists.txt
===================================================================
--- lib/Target/MSP430/CMakeLists.txt
+++ lib/Target/MSP430/CMakeLists.txt
@@ -20,7 +20,19 @@
   MSP430TargetMachine.cpp
   MSP430AsmPrinter.cpp
   MSP430MCInstLower.cpp
-  )
+
+  LINK_COMPONENTS
+    AsmPrinter
+    CodeGen
+    Core
+    MC
+    MSP430AsmPrinter
+    MSP430Desc
+    MSP430Info
+    SelectionDAG
+    Support
+    Target
+)
 
 add_subdirectory(InstPrinter)
 add_subdirectory(TargetInfo)
Index: lib/Target/MSP430/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/MSP430/InstPrinter/CMakeLists.txt
+++ lib/Target/MSP430/InstPrinter/CMakeLists.txt
@@ -1,3 +1,7 @@
 add_llvm_library(LLVMMSP430AsmPrinter
   MSP430InstPrinter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+)
Index: lib/Target/MSP430/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/MSP430/MCTargetDesc/CMakeLists.txt
+++ lib/Target/MSP430/MCTargetDesc/CMakeLists.txt
@@ -1,4 +1,10 @@
 add_llvm_library(LLVMMSP430Desc
   MSP430MCTargetDesc.cpp
   MSP430MCAsmInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    MSP430AsmPrinter
+    MSP430Info
+    Support
+)
Index: lib/Target/MSP430/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/MSP430/TargetInfo/CMakeLists.txt
+++ lib/Target/MSP430/TargetInfo/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMMSP430Info
   MSP430TargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/Mips/AsmParser/CMakeLists.txt
===================================================================
--- lib/Target/Mips/AsmParser/CMakeLists.txt
+++ lib/Target/Mips/AsmParser/CMakeLists.txt
@@ -1,3 +1,10 @@
 add_llvm_library(LLVMMipsAsmParser
   MipsAsmParser.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    MCParser
+    MipsDesc
+    MipsInfo
+    Support
+)
Index: lib/Target/Mips/CMakeLists.txt
===================================================================
--- lib/Target/Mips/CMakeLists.txt
+++ lib/Target/Mips/CMakeLists.txt
@@ -47,7 +47,20 @@
   MipsSubtarget.cpp
   MipsTargetMachine.cpp
   MipsTargetObjectFile.cpp
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    AsmPrinter
+    CodeGen
+    Core
+    MC
+    MipsAsmPrinter
+    MipsDesc
+    MipsInfo
+    SelectionDAG
+    Support
+    Target
+)
 
 add_subdirectory(InstPrinter)
 add_subdirectory(Disassembler)
Index: lib/Target/Mips/Disassembler/CMakeLists.txt
===================================================================
--- lib/Target/Mips/Disassembler/CMakeLists.txt
+++ lib/Target/Mips/Disassembler/CMakeLists.txt
@@ -1,3 +1,8 @@
 add_llvm_library(LLVMMipsDisassembler
   MipsDisassembler.cpp
-  )
+
+  LINK_COMPONENTS
+    MCDisassembler
+    MipsInfo
+    Support
+)
Index: lib/Target/Mips/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/Mips/InstPrinter/CMakeLists.txt
+++ lib/Target/Mips/InstPrinter/CMakeLists.txt
@@ -1,3 +1,7 @@
 add_llvm_library(LLVMMipsAsmPrinter
   MipsInstPrinter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+)
Index: lib/Target/Mips/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/Mips/MCTargetDesc/CMakeLists.txt
+++ lib/Target/Mips/MCTargetDesc/CMakeLists.txt
@@ -11,4 +11,10 @@
   MipsNaClELFStreamer.cpp
   MipsOptionRecord.cpp
   MipsTargetStreamer.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    MipsAsmPrinter
+    MipsInfo
+    Support
+)
Index: lib/Target/Mips/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/Mips/TargetInfo/CMakeLists.txt
+++ lib/Target/Mips/TargetInfo/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMMipsInfo
   MipsTargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/NVPTX/CMakeLists.txt
===================================================================
--- lib/Target/NVPTX/CMakeLists.txt
+++ lib/Target/NVPTX/CMakeLists.txt
@@ -35,7 +35,24 @@
   NVVMReflect.cpp
   )
 
-add_llvm_target(NVPTXCodeGen ${NVPTXCodeGen_sources})
+add_llvm_target(NVPTXCodeGen ${NVPTXCodeGen_sources}
+
+  LINK_COMPONENTS
+    Analysis
+    AsmPrinter
+    CodeGen
+    Core
+    MC
+    NVPTXAsmPrinter
+    NVPTXDesc
+    NVPTXInfo
+    ScalarOpts
+    SelectionDAG
+    Support
+    Target
+    TransformUtils
+    Vectorize
+)
 
 add_subdirectory(TargetInfo)
 add_subdirectory(InstPrinter)
Index: lib/Target/NVPTX/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/NVPTX/InstPrinter/CMakeLists.txt
+++ lib/Target/NVPTX/InstPrinter/CMakeLists.txt
@@ -1,3 +1,7 @@
 add_llvm_library(LLVMNVPTXAsmPrinter
   NVPTXInstPrinter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+)
Index: lib/Target/NVPTX/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/NVPTX/MCTargetDesc/CMakeLists.txt
+++ lib/Target/NVPTX/MCTargetDesc/CMakeLists.txt
@@ -1,4 +1,10 @@
 add_llvm_library(LLVMNVPTXDesc
   NVPTXMCAsmInfo.cpp
   NVPTXMCTargetDesc.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    NVPTXAsmPrinter
+    NVPTXInfo
+    Support
+)
Index: lib/Target/NVPTX/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/NVPTX/TargetInfo/CMakeLists.txt
+++ lib/Target/NVPTX/TargetInfo/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMNVPTXInfo
   NVPTXTargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/PowerPC/AsmParser/CMakeLists.txt
===================================================================
--- lib/Target/PowerPC/AsmParser/CMakeLists.txt
+++ lib/Target/PowerPC/AsmParser/CMakeLists.txt
@@ -1,3 +1,10 @@
 add_llvm_library(LLVMPowerPCAsmParser
   PPCAsmParser.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    MCParser
+    PowerPCDesc
+    PowerPCInfo
+    Support
+)
Index: lib/Target/PowerPC/CMakeLists.txt
===================================================================
--- lib/Target/PowerPC/CMakeLists.txt
+++ lib/Target/PowerPC/CMakeLists.txt
@@ -40,7 +40,22 @@
   PPCVSXCopy.cpp
   PPCVSXFMAMutate.cpp
   PPCVSXSwapRemoval.cpp
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    AsmPrinter
+    CodeGen
+    Core
+    MC
+    PowerPCAsmPrinter
+    PowerPCDesc
+    PowerPCInfo
+    ScalarOpts
+    SelectionDAG
+    Support
+    Target
+    TransformUtils
+)
 
 add_subdirectory(AsmParser)
 add_subdirectory(Disassembler)
Index: lib/Target/PowerPC/Disassembler/CMakeLists.txt
===================================================================
--- lib/Target/PowerPC/Disassembler/CMakeLists.txt
+++ lib/Target/PowerPC/Disassembler/CMakeLists.txt
@@ -1,3 +1,8 @@
 add_llvm_library(LLVMPowerPCDisassembler
   PPCDisassembler.cpp
-  )
+
+  LINK_COMPONENTS
+    MCDisassembler
+    PowerPCInfo
+    Support
+)
Index: lib/Target/PowerPC/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/PowerPC/InstPrinter/CMakeLists.txt
+++ lib/Target/PowerPC/InstPrinter/CMakeLists.txt
@@ -1,3 +1,7 @@
 add_llvm_library(LLVMPowerPCAsmPrinter
   PPCInstPrinter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+)
Index: lib/Target/PowerPC/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/PowerPC/MCTargetDesc/CMakeLists.txt
+++ lib/Target/PowerPC/MCTargetDesc/CMakeLists.txt
@@ -7,4 +7,10 @@
   PPCPredicates.cpp
   PPCMachObjectWriter.cpp
   PPCELFObjectWriter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    PowerPCAsmPrinter
+    PowerPCInfo
+    Support
+)
Index: lib/Target/PowerPC/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/PowerPC/TargetInfo/CMakeLists.txt
+++ lib/Target/PowerPC/TargetInfo/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMPowerPCInfo
   PowerPCTargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/RISCV/CMakeLists.txt
===================================================================
--- lib/Target/RISCV/CMakeLists.txt
+++ lib/Target/RISCV/CMakeLists.txt
@@ -8,7 +8,14 @@
 
 add_llvm_target(RISCVCodeGen
   RISCVTargetMachine.cpp
-  )
+
+  LINK_COMPONENTS
+    Core
+    CodeGen
+    RISCVInfo
+    Support
+    Target
+)
 
 add_subdirectory(TargetInfo)
 add_subdirectory(MCTargetDesc)
Index: lib/Target/RISCV/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/RISCV/MCTargetDesc/CMakeLists.txt
+++ lib/Target/RISCV/MCTargetDesc/CMakeLists.txt
@@ -4,4 +4,9 @@
   RISCVMCAsmInfo.cpp
   RISCVMCTargetDesc.cpp
   RISCVMCCodeEmitter.cpp
+
+  LINK_COMPONENTS
+    MC
+    RISCVInfo
+    Support
 )
Index: lib/Target/RISCV/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/RISCV/TargetInfo/CMakeLists.txt
+++ lib/Target/RISCV/TargetInfo/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMRISCVInfo
   RISCVTargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/Sparc/AsmParser/CMakeLists.txt
===================================================================
--- lib/Target/Sparc/AsmParser/CMakeLists.txt
+++ lib/Target/Sparc/AsmParser/CMakeLists.txt
@@ -1,3 +1,10 @@
 add_llvm_library(LLVMSparcAsmParser
   SparcAsmParser.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    MCParser
+    SparcDesc
+    SparcInfo
+    Support
+)
Index: lib/Target/Sparc/CMakeLists.txt
===================================================================
--- lib/Target/Sparc/CMakeLists.txt
+++ lib/Target/Sparc/CMakeLists.txt
@@ -25,7 +25,19 @@
   SparcTargetMachine.cpp
   SparcMCInstLower.cpp
   SparcTargetObjectFile.cpp
-  )
+
+  LINK_COMPONENTS
+    AsmPrinter
+    CodeGen
+    Core
+    MC
+    SelectionDAG
+    SparcAsmPrinter
+    SparcDesc
+    SparcInfo
+    Support
+    Target
+)
 
 add_subdirectory(TargetInfo)
 add_subdirectory(MCTargetDesc)
Index: lib/Target/Sparc/Disassembler/CMakeLists.txt
===================================================================
--- lib/Target/Sparc/Disassembler/CMakeLists.txt
+++ lib/Target/Sparc/Disassembler/CMakeLists.txt
@@ -1,3 +1,8 @@
 add_llvm_library(LLVMSparcDisassembler
   SparcDisassembler.cpp
-  )
+
+  LINK_COMPONENTS
+    MCDisassembler
+    SparcInfo
+    Support
+)
Index: lib/Target/Sparc/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/Sparc/InstPrinter/CMakeLists.txt
+++ lib/Target/Sparc/InstPrinter/CMakeLists.txt
@@ -1,3 +1,7 @@
 add_llvm_library(LLVMSparcAsmPrinter
   SparcInstPrinter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+)
Index: lib/Target/Sparc/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/Sparc/MCTargetDesc/CMakeLists.txt
+++ lib/Target/Sparc/MCTargetDesc/CMakeLists.txt
@@ -6,4 +6,10 @@
   SparcMCTargetDesc.cpp
   SparcMCExpr.cpp
   SparcTargetStreamer.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    SparcAsmPrinter
+    SparcInfo
+    Support
+)
Index: lib/Target/Sparc/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/Sparc/TargetInfo/CMakeLists.txt
+++ lib/Target/Sparc/TargetInfo/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMSparcInfo
   SparcTargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/SystemZ/AsmParser/CMakeLists.txt
===================================================================
--- lib/Target/SystemZ/AsmParser/CMakeLists.txt
+++ lib/Target/SystemZ/AsmParser/CMakeLists.txt
@@ -1,3 +1,10 @@
 add_llvm_library(LLVMSystemZAsmParser
   SystemZAsmParser.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    MCParser
+    Support
+    SystemZDesc
+    SystemZInfo
+)
Index: lib/Target/SystemZ/CMakeLists.txt
===================================================================
--- lib/Target/SystemZ/CMakeLists.txt
+++ lib/Target/SystemZ/CMakeLists.txt
@@ -34,7 +34,20 @@
   SystemZTargetMachine.cpp
   SystemZTargetTransformInfo.cpp
   SystemZTDC.cpp
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    AsmPrinter
+    CodeGen
+    Core
+    MC
+    SelectionDAG
+    Support
+    SystemZAsmPrinter
+    SystemZDesc
+    SystemZInfo
+    Target
+)
 
 add_subdirectory(AsmParser)
 add_subdirectory(Disassembler)
Index: lib/Target/SystemZ/Disassembler/CMakeLists.txt
===================================================================
--- lib/Target/SystemZ/Disassembler/CMakeLists.txt
+++ lib/Target/SystemZ/Disassembler/CMakeLists.txt
@@ -1,3 +1,10 @@
 add_llvm_library(LLVMSystemZDisassembler
   SystemZDisassembler.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    MCDisassembler
+    Support
+    SystemZDesc
+    SystemZInfo
+)
Index: lib/Target/SystemZ/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/SystemZ/InstPrinter/CMakeLists.txt
+++ lib/Target/SystemZ/InstPrinter/CMakeLists.txt
@@ -1,3 +1,7 @@
 add_llvm_library(LLVMSystemZAsmPrinter
   SystemZInstPrinter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+)
Index: lib/Target/SystemZ/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/SystemZ/MCTargetDesc/CMakeLists.txt
+++ lib/Target/SystemZ/MCTargetDesc/CMakeLists.txt
@@ -4,4 +4,10 @@
   SystemZMCCodeEmitter.cpp
   SystemZMCObjectWriter.cpp
   SystemZMCTargetDesc.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+    SystemZAsmPrinter
+    SystemZInfo
+)
Index: lib/Target/SystemZ/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/SystemZ/TargetInfo/CMakeLists.txt
+++ lib/Target/SystemZ/TargetInfo/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMSystemZInfo
   SystemZTargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/WebAssembly/CMakeLists.txt
===================================================================
--- lib/Target/WebAssembly/CMakeLists.txt
+++ lib/Target/WebAssembly/CMakeLists.txt
@@ -46,6 +46,21 @@
 
   DEPENDS
   intrinsics_gen
+
+  LINK_COMPONENTS
+    Analysis
+    AsmPrinter
+    CodeGen
+    Core
+    MC
+    ScalarOpts
+    SelectionDAG
+    Support
+    Target
+    TransformUtils
+    WebAssemblyAsmPrinter
+    WebAssemblyDesc
+    WebAssemblyInfo
 )
 
 add_subdirectory(Disassembler)
Index: lib/Target/WebAssembly/Disassembler/CMakeLists.txt
===================================================================
--- lib/Target/WebAssembly/Disassembler/CMakeLists.txt
+++ lib/Target/WebAssembly/Disassembler/CMakeLists.txt
@@ -1,3 +1,8 @@
 add_llvm_library(LLVMWebAssemblyDisassembler
   WebAssemblyDisassembler.cpp
-  )
+
+  LINK_COMPONENTS
+    MCDisassembler
+    WebAssemblyInfo
+    Support
+)
Index: lib/Target/WebAssembly/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/WebAssembly/InstPrinter/CMakeLists.txt
+++ lib/Target/WebAssembly/InstPrinter/CMakeLists.txt
@@ -1,3 +1,7 @@
 add_llvm_library(LLVMWebAssemblyAsmPrinter
   WebAssemblyInstPrinter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+)
Index: lib/Target/WebAssembly/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/WebAssembly/MCTargetDesc/CMakeLists.txt
+++ lib/Target/WebAssembly/MCTargetDesc/CMakeLists.txt
@@ -5,4 +5,10 @@
   WebAssemblyMCCodeEmitter.cpp
   WebAssemblyMCTargetDesc.cpp
   WebAssemblyTargetStreamer.cpp
+
+  LINK_COMPONENTS
+    MC
+    Support
+    WebAssemblyAsmPrinter
+    WebAssemblyInfo
 )
Index: lib/Target/WebAssembly/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/WebAssembly/TargetInfo/CMakeLists.txt
+++ lib/Target/WebAssembly/TargetInfo/CMakeLists.txt
@@ -2,6 +2,9 @@
 
 add_llvm_library(LLVMWebAssemblyInfo
   WebAssemblyTargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
 
 add_dependencies(LLVMWebAssemblyInfo WebAssemblyCommonTableGen)
Index: lib/Target/X86/AsmParser/CMakeLists.txt
===================================================================
--- lib/Target/X86/AsmParser/CMakeLists.txt
+++ lib/Target/X86/AsmParser/CMakeLists.txt
@@ -1,4 +1,11 @@
 add_llvm_library(LLVMX86AsmParser
   X86AsmInstrumentation.cpp
   X86AsmParser.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    MCParser
+    Support
+    X86Desc
+    X86Info
+)
Index: lib/Target/X86/CMakeLists.txt
===================================================================
--- lib/Target/X86/CMakeLists.txt
+++ lib/Target/X86/CMakeLists.txt
@@ -59,7 +59,23 @@
   ${GLOBAL_ISEL_BUILD_FILES}
   )
 
-add_llvm_target(X86CodeGen ${sources})
+add_llvm_target(X86CodeGen ${sources}
+
+  LINK_COMPONENTS
+    Analysis
+    AsmPrinter
+    CodeGen
+    Core
+    MC
+    SelectionDAG
+    Support
+    Target
+    X86AsmPrinter
+    X86Desc
+    X86Info
+    X86Utils
+    GlobalISel
+)
 
 add_subdirectory(AsmParser)
 add_subdirectory(Disassembler)
Index: lib/Target/X86/Disassembler/CMakeLists.txt
===================================================================
--- lib/Target/X86/Disassembler/CMakeLists.txt
+++ lib/Target/X86/Disassembler/CMakeLists.txt
@@ -1,4 +1,9 @@
 add_llvm_library(LLVMX86Disassembler
   X86Disassembler.cpp
   X86DisassemblerDecoder.cpp
-  )
+
+  LINK_COMPONENTS
+    MCDisassembler
+    Support
+    X86Info
+)
Index: lib/Target/X86/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/X86/InstPrinter/CMakeLists.txt
+++ lib/Target/X86/InstPrinter/CMakeLists.txt
@@ -2,4 +2,9 @@
   X86ATTInstPrinter.cpp
   X86IntelInstPrinter.cpp
   X86InstComments.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+    X86Utils
+)
Index: lib/Target/X86/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/X86/MCTargetDesc/CMakeLists.txt
+++ lib/Target/X86/MCTargetDesc/CMakeLists.txt
@@ -7,4 +7,12 @@
   X86ELFObjectWriter.cpp
   X86WinCOFFStreamer.cpp
   X86WinCOFFObjectWriter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    MCDisassembler
+    Object
+    Support
+    X86AsmPrinter
+    X86Info
+)
Index: lib/Target/X86/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/X86/TargetInfo/CMakeLists.txt
+++ lib/Target/X86/TargetInfo/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMX86Info
   X86TargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Target/X86/Utils/CMakeLists.txt
===================================================================
--- lib/Target/X86/Utils/CMakeLists.txt
+++ lib/Target/X86/Utils/CMakeLists.txt
@@ -1,3 +1,7 @@
 add_llvm_library(LLVMX86Utils
   X86ShuffleDecode.cpp
-  )
+
+  LINK_COMPONENTS
+    Core
+    Support
+)
Index: lib/Target/XCore/CMakeLists.txt
===================================================================
--- lib/Target/XCore/CMakeLists.txt
+++ lib/Target/XCore/CMakeLists.txt
@@ -24,7 +24,21 @@
   XCoreTargetObjectFile.cpp
   XCoreSelectionDAGInfo.cpp
   XCoreFrameToArgsOffsetElim.cpp
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    AsmPrinter
+    CodeGen
+    Core
+    MC
+    SelectionDAG
+    Support
+    Target
+    TransformUtils
+    XCoreAsmPrinter
+    XCoreDesc
+    XCoreInfo
+)
 
 add_subdirectory(Disassembler)
 add_subdirectory(InstPrinter)
Index: lib/Target/XCore/Disassembler/CMakeLists.txt
===================================================================
--- lib/Target/XCore/Disassembler/CMakeLists.txt
+++ lib/Target/XCore/Disassembler/CMakeLists.txt
@@ -1,3 +1,8 @@
 add_llvm_library(LLVMXCoreDisassembler
   XCoreDisassembler.cpp
-  )
+
+  LINK_COMPONENTS
+    MCDisassembler
+    Support
+    XCoreInfo
+)
Index: lib/Target/XCore/InstPrinter/CMakeLists.txt
===================================================================
--- lib/Target/XCore/InstPrinter/CMakeLists.txt
+++ lib/Target/XCore/InstPrinter/CMakeLists.txt
@@ -1,3 +1,7 @@
 add_llvm_library(LLVMXCoreAsmPrinter
   XCoreInstPrinter.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+)
Index: lib/Target/XCore/MCTargetDesc/CMakeLists.txt
===================================================================
--- lib/Target/XCore/MCTargetDesc/CMakeLists.txt
+++ lib/Target/XCore/MCTargetDesc/CMakeLists.txt
@@ -1,4 +1,10 @@
 add_llvm_library(LLVMXCoreDesc
   XCoreMCTargetDesc.cpp
   XCoreMCAsmInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    MC
+    Support
+    XCoreAsmPrinter
+    XCoreInfo
+)
Index: lib/Target/XCore/TargetInfo/CMakeLists.txt
===================================================================
--- lib/Target/XCore/TargetInfo/CMakeLists.txt
+++ lib/Target/XCore/TargetInfo/CMakeLists.txt
@@ -1,3 +1,6 @@
 add_llvm_library(LLVMXCoreInfo
   XCoreTargetInfo.cpp
-  )
+
+  LINK_COMPONENTS
+    Support
+)
Index: lib/Transforms/Coroutines/CMakeLists.txt
===================================================================
--- lib/Transforms/Coroutines/CMakeLists.txt
+++ lib/Transforms/Coroutines/CMakeLists.txt
@@ -8,4 +8,12 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    Core
+    ipo
+    ScalarOpts
+    Support
+    TransformUtils
+)
Index: lib/Transforms/IPO/CMakeLists.txt
===================================================================
--- lib/Transforms/IPO/CMakeLists.txt
+++ lib/Transforms/IPO/CMakeLists.txt
@@ -37,4 +37,19 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    BitWriter
+    Core
+    InstCombine
+    IRReader
+    Linker
+    Object
+    ProfileData
+    ScalarOpts
+    Support
+    TransformUtils
+    Vectorize
+    Instrumentation
+)
Index: lib/Transforms/InstCombine/CMakeLists.txt
===================================================================
--- lib/Transforms/InstCombine/CMakeLists.txt
+++ lib/Transforms/InstCombine/CMakeLists.txt
@@ -19,4 +19,10 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    Core
+    Support
+    TransformUtils
+)
Index: lib/Transforms/Instrumentation/CMakeLists.txt
===================================================================
--- lib/Transforms/Instrumentation/CMakeLists.txt
+++ lib/Transforms/Instrumentation/CMakeLists.txt
@@ -17,4 +17,12 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    Core
+    MC
+    Support
+    TransformUtils
+    ProfileData
+)
Index: lib/Transforms/ObjCARC/CMakeLists.txt
===================================================================
--- lib/Transforms/ObjCARC/CMakeLists.txt
+++ lib/Transforms/ObjCARC/CMakeLists.txt
@@ -14,4 +14,10 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    Core
+    Support
+    TransformUtils
+)
Index: lib/Transforms/Scalar/CMakeLists.txt
===================================================================
--- lib/Transforms/Scalar/CMakeLists.txt
+++ lib/Transforms/Scalar/CMakeLists.txt
@@ -63,4 +63,11 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    Core
+    InstCombine
+    Support
+    TransformUtils
+)
Index: lib/Transforms/Utils/CMakeLists.txt
===================================================================
--- lib/Transforms/Utils/CMakeLists.txt
+++ lib/Transforms/Utils/CMakeLists.txt
@@ -58,4 +58,9 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    Core
+    Support
+)
Index: lib/Transforms/Vectorize/CMakeLists.txt
===================================================================
--- lib/Transforms/Vectorize/CMakeLists.txt
+++ lib/Transforms/Vectorize/CMakeLists.txt
@@ -10,4 +10,10 @@
 
   DEPENDS
   intrinsics_gen
-  )
+
+  LINK_COMPONENTS
+    Analysis
+    Core
+    Support
+    TransformUtils
+)
